thoth/storages/result_base.py

```diff
@@ -56,9 +57,20 @@ def __init__(
         ), "Make sure RESULT_TYPE in derived classes to distinguish between adapter type instances is non-empty."
 
         self.deployment_name = deployment_name or os.environ["THOTH_DEPLOYMENT_NAME"]
-        self.prefix = "{}/{}/{}".format(
-            prefix or os.environ["THOTH_CEPH_BUCKET_PREFIX"], self.deployment_name, self.RESULT_TYPE
-        )
+
+        self.document_id_prefix = document_id_prefix or os.environ["THOTH_FILE_ID_PREFIX"]
```

Comment:
Can we have just a parameter to distinguish based on the prefix? It sounds to me misleading to provide an environment variable to adjust the document retrieval - a function parameter can be more suited. The parameter can be directly `datetime` so we have a comfortable API to query the documents.

Reply:
> Can we have just a parameter to distinguish based on the prefix? It sounds to me misleading to provide an environment variable to adjust the document retrieval - a function parameter can be more suited. The parameter can be directly `datetime` so we have a comfortable API to query the documents.

what about `file_id_prefix` parameter? not to constraint only to datetime?

Reply:
updated in https://github.com/thoth-station/storages/pull/2273/commits/00a58740a8bac5b672d44bcc7841c9b91e458ec0

Reply:
Won't the prefix always be `<doctype>-<datetime>-<hash>`?

Reply:
Actually yes, but `CephStore` class is general not only for documents, so maybe for `ResultStorageBase` makes sense to have `datetime` as document ids are recreated like that in common.

Reply:
updated in https://github.com/thoth-station/storages/pull/2273/commits/65343ef3711789d7494577baea621f20d58ac635

Reply:
```
import datetime
from thoth.storages import AdvisersResultsStore

adviser = AdvisersResultsStore(
    datetime_= datetime.datetime.utcnow() - datetime.timedelta(days=50)
)

adviser.connect()

for idd in adviser.get_document_listing():
    print(idd)

[fmurdaca@pc-7 16:47:08 ~/work/aicoe/storages](introduce-document-id-prefix)$ pipenv run python3 test_ceph.py 
210216
adviser-210216182351-6ccbc250e532f9f8
adviser-210216182958-c578f62785eb8761
adviser-210216200858-1b2480815dc96a69
adviser-210216223227-ce381787ded17b0a
adviser-210216223307-577727b37bb17f7d
adviser-210216223403-de9afa93319b716b
adviser-210226140918-610eef7a71c7cb04
adviser-210226142440-503749bba97b3fee
adviser-210226232050-4032b7921a0aa843
```

Reply:
I don't want to nitpick here but I think the design could be improved. I don't see a reason why the datetime object should be passed to the adapter constructor. Rather, I would like to instantiate the adapter instance once and parametrize the operation that should be performed:

```python
import datetime, date, timedelta
from thoth.storages import AdvisersResultsStore

adviser = AdvisersResultsStore()
adviser.connect()

#
# All the documents available.
#
for idd in adviser.get_document_listing(): 
    print(idd)

#
# All the documents created in the past 30 days.
#
datetime_last30 = datetime.datetime.utcnow() - datetime.timedelta(days=30)
for idd in adviser.get_document_listing(start_datetime=datetime_last30):
    print(idd)

#
# All the documents created last month.
#
last_month_start = date.today().replace(day=1) - timedelta(days=1)
last_month_end = date.today().replace(day=1) - timedelta(days=last_month_start.day)
for idd in adviser.get_document_listing(start_datetime=datetime_, end_datetime=last_month_end):
    print(idd)
```

Such design unlocks the possibility to do analytics on the use - I can imagine for example a bar chart showing how the system is utilized based on day of week, based on months or based on hour that could be nice data analyses to show.

Reply:
updated in https://github.com/thoth-station/storages/pull/2273/commits/b71440ba3b8c36b08d3770d0f306306ad0ea071c, but not sure how we can provide two filters at once using prefix unless we take all document ids and verify which of them is within a certain range. 

current solution:
```
def get_document_listing(self, datetime_: typing.Optional[datetime.datetime] = None) -> typing.Generator[str, None, None]:
        """Get listing of documents stored on the Ceph."""
        prefix = self.prefix + f"{self.prefix.split('/')[-2]}-{datetime_:%y%m%d}" if datetime_ else self.prefix
        for obj in self._s3.Bucket(self.bucket).objects.filter(Prefix=prefix).all():
            yield obj.key[len(prefix) :]  # Ignore PycodestyleBear (E203)
```

Reply:
> way better, thanks @fridex, updated in [b71440b](https://github.com/thoth-station/storages/commit/b71440ba3b8c36b08d3770d0f306306ad0ea071c),
> 
> ```
> import datetime
> from thoth.storages import AdvisersResultsStore
> 
> datetime_ = datetime.datetime.utcnow() - datetime.timedelta(days=51)
> adviser = AdvisersResultsStore()
> 
> adviser.connect()
> 
> for idd in adviser.get_document_listing(datetime_=datetime_):
>     print(idd)
> ```
> 
> but not sure how we can provide two filters at once using prefix unless we take all document ids and verify which of them is within a certain range. I did not find a filter by datetime in boto3 for this case. maybe a feature request?
> 
> current solution:
> 

I generalize this logic to be `filter_` from CephStore class and placed the `datetime_` to ResultStorageBase class because it is more specialized case for thoth document naming convention. See: https://github.com/thoth-station/storages/pull/2273/commits/ba2b43c9d476344c623a12da3f3e811cd3e4bcfc
> ```
>     def get_document_listing(self, datetime_: typing.Optional[datetime.datetime] = None) -> typing.Generator[str, None, None]:
>         """Get listing of documents stored on the Ceph."""
>         prefix = self.prefix + f"{self.prefix.split('/')[-2]}-{datetime_:%y%m%d}" if datetime_ else self.prefix
>         for obj in self._s3.Bucket(self.bucket).objects.filter(Prefix=prefix).all():
>             yield obj.key[len(prefix) :]  # Ignore PycodestyleBear (E203)
> ```

Reply:
> I don't want to nitpick here but I think the design could be improved. I don't see a reason why the datetime object should be passed to the adapter constructor. Rather, I would like to instantiate the adapter instance once and parametrize the operation that should be performed:
> 
> ```python
> import datetime, date, timedelta
> from thoth.storages import AdvisersResultsStore
> 
> adviser = AdvisersResultsStore()
> adviser.connect()
> 
> #
> # All the documents available.
> #
> for idd in adviser.get_document_listing(): 
>     print(idd)
> 
> #
> # All the documents created in the past 30 days.
> #
> datetime_last30 = datetime.datetime.utcnow() - datetime.timedelta(days=30)
> for idd in adviser.get_document_listing(start_datetime=datetime_last30):
>     print(idd)

```python
import datetime, date, timedelta
from thoth.storages import AdvisersResultsStore

adviser = AdvisersResultsStore()
adviser.connect()
```
with the current approach you can do:

```python
for day in range(1, 30):
    datetime_ = date.today() - timedelta(days=day)
    for idd in adviser.get_document_listing(datetime_=datetime_):
          print(idd)
```

> 
> #
> # All the documents created last month.
> #
> last_month_start = date.today().replace(day=1) - timedelta(days=1)
> last_month_end = date.today().replace(day=1) - timedelta(days=last_month_start.day)
> for idd in adviser.get_document_listing(start_datetime=last_month_start, end_datetime=last_month_end):
>     print(idd)

with the current approach you can do:

```python
for day in range(1, 30):
    datetime_ = date.today() - timedelta(days=day)
    for idd in adviser.get_document_listing(datetime_=datetime_):
          print(idd)
```

> 
> Such design unlocks the possibility to do analytics on the use - I can imagine for example a bar chart showing how the system is utilized based on day of week, based on months or based on hour that could be nice data analyses to show.